# LIBRARY

In [41]:
!pip install miceforest

In [42]:
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer, KNNImputer
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
import re
import miceforest as mf
import warnings
warnings.filterwarnings("ignore")

# LOAD DATASET

In [43]:
features = pd.read_csv("/content/features.csv")
labels = pd.read_csv("/content/labels.csv")

# 1️. Tampilkan data yang telah diunduh, kemudian jelaskan semua kolom yang terdapat pada dataset tersebut.


In [44]:
features.head().T

0           1  \
Unnamed: 0                                               0           1   
Student Id                                            6634        6459   
Marital status                                      single      single   
Application mode                                       1.0         1.0   
Application order                                      3.0         1.0   
Course                                              9773.0         NaN   
Daytime/evening attendance\t                           day         day   
Previous qualification                                 1.0         1.0   
Previous qualification (grade)                         NaN       136.0   
Nacionality                                     Portuguese  Portuguese   
Mother's qualification                                 1.0        19.0   
Father's qualification                                 1.0        19.0   
Mother's occupation                                    4.0         4.0   
Father's occupation                                    7.0         8.0   
Admission grade                                      125.1       150.4   
Displaced                                              1.0         1.0   
Educational special needs                              0.0         0.0   
Debtor                                                 1.0         0.0   
Tuition fees up to date                                1.0         1.0   
Gender                                                 1.0         1.0   
Scholarship holder                                     NaN         1.0   
Age at enrollment                                     18.0        21.0   
International                                          0.0         0.0   
Curricular units 1st sem (credited)                    0.0         0.0   
Curricular units 1st sem (enrolled)                    6.0         0.0   
Curricular units 1st sem (evaluations)                 6.0         0.0   
Curricular units 1st sem (approved)                    6.0         0.0   
Curricular units 1st sem (grade)                 14.333333         0.0   
Curricular units 1st sem (without evaluations)         0.0         0.0   
Curricular units 2nd sem (credited)                    0.0         0.0   
Curricular units 2nd sem (enrolled)                    6.0         0.0   
Curricular units 2nd sem (evaluations)                 6.0         0.0   
Curricular units 2nd sem (approved)                    6.0         0.0   
Curricular units 2nd sem (grade)                 15.333333         0.0   
Curricular units 2nd sem (without evaluations)         0.0         0.0   
Unemployment rate                                     11.1        12.4   
Inflation rate                                         0.6         0.5   
GDP                                                   2.02        1.79   

                                                         2           3  \
Unnamed: 0                                               2           3   
Student Id                                            2238        2479   
Marital status                                      single      single   
Application mode                                      42.0         1.0   
Application order                                      1.0         1.0   
Course                                              9500.0         NaN   
Daytime/evening attendance\t                           day         day   
Previous qualification                                 1.0         1.0   
Previous qualification (grade)                       120.0       141.0   
Nacionality                                     Portuguese  Portuguese   
Mother's qualification                                 1.0         1.0   
Father's qualification                                 2.0        19.0   
Mother's occupation                                    9.0         4.0   
Father's occupation                                    3.0         4.0   
Admission grade                                      130.0  

In [45]:
labels.head()

Target
0  Graduate
1  Graduate
2  Graduate
3  Graduate
4   Dropout

PENJELASAN KOLOM DATASET:

1. Student Id : ID unik mahasiswa.
2. Marital status : Status pernikahan mahasiswa (single, married, dll).
3. Application mode : Cara atau jalur pendaftaran mahasiswa.
4. Application order : Urutan pilihan program studi yang diajukan saat mendaftar.
5. Course : Kode program studi (misalnya 9773 = Informatika, 9500 = Manajemen, dll).
6. Daytime/evening attendance : Jenis kelas (day = reguler pagi, evening = kelas malam).
7. Previous qualification : Jenis kualifikasi pendidikan sebelumnya (misal SMA, diploma).
8. Previous qualification (grade) : Nilai akhir dari pendidikan sebelumnya.
9. Nacionality : Kewarganegaraan mahasiswa.
10. Mother's qualification : Pendidikan terakhir ibu.
11. Father's qualification : Pendidikan terakhir ayah.
12. Mother's occupation : Pekerjaan ibu.
13. Father's occupation : Pekerjaan ayah.
14. Admission grade : Nilai ujian masuk universitas.
15. Displaced : Apakah mahasiswa tinggal jauh dari keluarga (1 = ya, 0 = tidak).
16. Educational special needs : Apakah mahasiswa memiliki kebutuhan khusus (1 = ya, 0 = tidak).
17. Debtor : Apakah mahasiswa memiliki tunggakan pembayaran (1 = ya, 0 = tidak).
18. Tuition fees up to date : Apakah pembayaran biaya kuliah sudah lunas.
19. Gender : Jenis kelamin mahasiswa (male/female).
20. Scholarship holder : Apakah mahasiswa penerima beasiswa.
21. Age at enrollment : Umur mahasiswa saat mendaftar kuliah.
22. International : Status mahasiswa internasional (1 = ya, 0 = tidak).
23. Curricular units 1st sem (credited) : Jumlah mata kuliah semester 1 yang diakui dari tempat lain.
24. Curricular units 1st sem (enrolled) : Jumlah mata kuliah semester 1 yang diambil.
25. Curricular units 1st sem (evaluations) : Jumlah evaluasi yang diikuti pada semester 1.
26. Curricular units 1st sem (approved) : Jumlah mata kuliah semester 1 yang lulus.
27. Curricular units 1st sem (grade) : Rata-rata nilai mata kuliah semester 1.
28. Curricular units 1st sem (without evaluations) : Jumlah mata kuliah semester 1 tanpa evaluasi.
29. Curricular units 2nd sem (credited) : Jumlah mata kuliah semester 2 yang diakui.
30. Curricular units 2nd sem (enrolled) : Jumlah mata kuliah semester 2 yang diambil.
31. Curricular units 2nd sem (evaluations) : Jumlah evaluasi yang diikuti semester 2.
32. Curricular units 2nd sem (approved) : Jumlah mata kuliah semester 2 yang lulus.
33. Curricular units 2nd sem (grade) : Rata-rata nilai semester 2.
34. Curricular units 2nd sem (without evaluations) : Jumlah mata kuliah semester 2 tanpa evaluasi.
35. Unemployment rate : Tingkat pengangguran nasional pada tahun tersebut.
36. Inflation rate : Tingkat inflasi nasional.
37. GDP : Pertumbuhan Produk Domestik Bruto (PDB) nasional.
38. Target : Label hasil (Graduate = lulus, Dropout = berhenti kuliah).

# 2. agregasi antara tabel features dengan tabel labels

In [46]:
df = pd.concat([features, labels], axis=1)
df

Unnamed: 0  Student Id Marital status  Application mode  \
0              0        6634         single               1.0   
1              1        6459         single               1.0   
2              2        2238         single              42.0   
3              3        2479         single               1.0   
4              4        4260         single              44.0   
...          ...         ...            ...               ...   
3091        3091        6176         single              44.0   
3092        3092        8064         single               1.0   
3093        3093        2103         single               1.0   
3094        3094        8629         single              39.0   
3095        3095        6614         single              17.0   

      Application order  Course Daytime/evening attendance\t  \
0                   3.0  9773.0                          day   
1                   1.0     NaN                          day   
2                   1.0  9500.0                          day   
3                   1.0     NaN                          day   
4                   1.0     NaN                          day   
...                 ...     ...                          ...   
3091                1.0  9085.0                          day   
3092                5.0  9238.0                          day   
3093                2.0  9500.0                          day   
3094                1.0  9500.0                          day   
3095                1.0  9670.0                          day   

      Previous qualification  Previous qualification (grade) Nacionality  ...  \
0                        1.0                             NaN  Portuguese  ...   
1                        1.0                           136.0  Portuguese  ...   
2                        1.0                           120.0  Portuguese  ...   
3                        1.0                           141.0  Portuguese  ...   
4                       39.0                           120.0     Spanish  ...   
...                      ...                             ...         ...  ...   
3091                     1.0                           140.0  Portuguese  ...   
3092                     1.0                           126.0  Portuguese  ...   
3093                     1.0                           141.0  Portuguese  ...   
3094                     4.0                           150.0  Portuguese  ...   
3095                     1.0                           118.0  Portuguese  ...   

      Curricular units 2nd sem (credited)  \
0                                     0.0   
1                                     0.0   
2                                     4.0   
3                                     0.0   
4                                     1.0   
...                                   ...   
3091                                  1.0   
3092                                  0.0   
3093                                  0.0   
3094                                  2.0   
3095                                  0.0   

      Curricular units 2nd sem (enrolled)  \
0                                     6.0   
1                                     0.0   
2                                     8.0   
3                                     6.0   
4                                     6.0   
...                                   ...   
3091                                  6.0   
3092                                  6.0   
3093                                  8.0   
3094                                  8.0   
3095                                  6.0   

      Curricular units 2nd sem (evaluations)  \
0                                        6.0   
1                                        0.0   
2                                        8.0   
3                                        7.0   
4                                        NaN   
...                                      ...   
3091                                     9.0   
3092             

# 3️. Tampilkan tipe data dari setiap kolom yang terdapat dalam dataset

In [47]:
df.dtypes #menampilkan tipe data

Unnamed: 0                                          int64
Student Id                                          int64
Marital status                                     object
Application mode                                  float64
Application order                                 float64
Course                                            float64
Daytime/evening attendance\t                       object
Previous qualification                            float64
Previous qualification (grade)                    float64
Nacionality                                        object
Mother's qualification                            float64
Father's qualification                            float64
Mother's occupation                               float64
Father's occupation                               float64
Admission grade                                   float64
Displaced                                         float64
Educational special needs                         float64
Debtor                                            float64
Tuition fees up to date                           float64
Gender                                            float64
Scholarship holder                                float64
Age at enrollment                                 float64
International                                     float64
Curricular units 1st sem (credited)               float64
Curricular units 1st sem (enrolled)               float64
Curricular units 1st sem (evaluations)            float64
Curricular units 1st sem (approved)               float64
Curricular units 1st sem (grade)                  float64
Curricular units 1st sem (without evaluations)    float64
Curricular units 2nd sem (credited)               float64
Curricular units 2nd sem (enrolled)               float64
Curricular units 2nd sem (evaluations)            float64
Curricular units 2nd sem (approved)               float64
Curricular units 2nd sem (grade)                  float64
Curricular units 2nd sem (without evaluations)    float64
Unemployment rate                                 float64
Inflation rate                                    float64
GDP                                               float64
Target                                             object
dtype: object

In [48]:
df.info() #menampilkan informasi data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3096 entries, 0 to 3095
Data columns (total 39 columns):
 #   Column                                          Non-Null Count  Dtype  
---  ------                                          --------------  -----  
 0   Unnamed: 0                                      3096 non-null   int64  
 1   Student Id                                      3096 non-null   int64  
 2   Marital status                                  3012 non-null   object 
 3   Application mode                                2924 non-null   float64
 4   Application order                               2988 non-null   float64
 5   Course                                          2944 non-null   float64
 6   Daytime/evening attendance	                     3016 non-null   object 
 7   Previous qualification                          2973 non-null   float64
 8   Previous qualification (grade)                  3011 non-null   float64
 9   Nacionality                              

# 4. Tampilkan statistik deskriptif dari dataset dan berikan analisis singkat mengenai hasilnya

In [49]:
df.describe().T

count         mean  \
Unnamed: 0                                      3096.0  1547.500000   
Student Id                                      3096.0  5470.623062   
Application mode                                2924.0    18.785568   
Application order                               2988.0     1.734940   
Course                                          2944.0  8817.929348   
Previous qualification                          2973.0     4.685503   
Previous qualification (grade)                  3011.0   132.647293   
Mother's qualification                          2951.0    19.470010   
Father's qualification                          3000.0    22.408667   
Mother's occupation                             3034.0    11.495715   
Father's occupation                             2952.0    11.597900   
Admission grade                                 2955.0   126.902876   
Displaced                                       2974.0     0.546066   
Educational special needs                       2977.0     0.010749   
Debtor                                          2980.0     0.117450   
Tuition fees up to date                         2940.0     0.880612   
Gender                                          2894.0     0.350726   
Scholarship holder                              2930.0     0.246416   
Age at enrollment                               2892.0    23.310166   
International                                   2995.0     0.025042   
Curricular units 1st sem (credited)             2888.0     0.698407   
Curricular units 1st sem (enrolled)             2953.0     6.251270   
Curricular units 1st sem (evaluations)          3029.0     8.233080   
Curricular units 1st sem (approved)             2912.0     4.732143   
Curricular units 1st sem (grade)                3007.0    10.692416   
Curricular units 1st sem (without evaluations)  3023.0     0.133311   
Curricular units 2nd sem (credited)             3035.0     0.516639   
Curricular units 2nd sem (enrolled)             2997.0     6.208876   
Curricular units 2nd sem (evaluations)          2889.0     8.061613   
Curricular units 2nd sem (approved)             2945.0     4.422411   
Curricular units 2nd sem (grade)                3021.0    10.275115   
Curricular units 2nd sem (without evaluations)  2943.0     0.139314   
Unemployment rate                               2960.0    11.549561   
Inflation rate                                  2929.0     1.215534   
GDP                                             3032.0     0.028278   

                                                        std      min      25%  \
Unnamed: 0                                       893.882543     0.00   773.75   
Student Id                                      2582.483019  1000.00  3288.50   
Application mode                                  17.532674     1.00     1.00   
Application order                                  1.327527     1.00     1.00   
Course                                          2148.407432    33.00  9085.00   
Previous qualification                            10.439709     1.00     1.00   
Previous qualification (grade)                    13.322662    95.00   124.00   
Mother's qualification                            15.636952     1.00     2.00   
Father's qualification                            15.291008     1.00     3.00   
Mother's occupation                               28.173386     0.00     4.00   
Father's occupation                               26.799023     0.00     4.00   
Admission grade                                   14.460283    95.00   118.00   
Displaced                                          0.497957     0.00     0.00   
Educational special needs                          0.103136     0.00     0.00   
Debtor                                             0.322009     0.00     0.00   
Tuition fees up to date                            0.324299     0.00     1.00   
Gender                                             0.477280     0.00     0.00   
Scholarship holder                

Analisis Singkat

  Profil Mahasiswa

Rata-rata usia mahasiswa saat pertama kali mendaftar adalah 23,3 tahun, dengan rentang dari 17 hingga 70 tahun. Ini berarti sebagian besar adalah mahasiswa usia tipikal, tapi tetap ada sejumlah kecil mahasiswa non-tradisional (dewasa atau lanjut usia) yang ikut menempuh pendidikan tinggi.

  Kinerja Akademik

Nilai masuk (Admission grade) rata-rata 126,9 dari skala 190, menandakan kualitas akademik calon mahasiswa tergolong menengah ke atas, meskipun variansinya cukup besar.

Mahasiswa rata-rata mengambil 6 mata kuliah per semester, tetapi hanya 4–5 yang berhasil lulus, menunjukkan bahwa tingkat kelulusan per semester sekitar 70–80%.

Nilai rata-rata per semester (grade) berada di kisaran 10–10,7, yang bisa dikategorikan sebagai kinerja akademik moderat, bukan buruk tapi jauh dari kategori unggul.

Ada sebagian kecil mahasiswa yang tidak menyelesaikan evaluasi semester (nilai = 0), kemungkinan besar berkaitan dengan absensi tinggi atau dropout di tengah semester.

  Kondisi Sosial dan Finansial

Sekitar 24,6% mahasiswa adalah penerima beasiswa, angka yang cukup signifikan dan menunjukkan adanya dukungan finansial yang kuat dari lembaga.

Hanya sekitar 11,7% mahasiswa yang memiliki tunggakan pembayaran (Debtor = 1), namun kelompok ini bisa jadi faktor utama dalam prediksi dropout karena tekanan finansial cenderung berdampak langsung pada kelanjutan studi.

Sebaliknya, 88% mahasiswa membayar biaya kuliah tepat waktu, yang merupakan indikator kedisiplinan dan stabilitas ekonomi yang baik.

  Kualitas Data

Sebagian besar kolom tidak memiliki data penuh — hampir setiap variabel memiliki nilai kosong (missing values) dengan tingkat kehilangan antara 3%–10%.
Artinya, sebelum data ini bisa digunakan untuk pemodelan prediktif, imputasi atau pembersihan data wajib dilakukan.
Khusus kolom akademik dan sosio-ekonomi (seperti “Parent’s qualification” dan “Admission grade”) perlu perhatian ekstra karena datanya sangat berpengaruh pada akurasi model.

# 5️. Slicing Data

In [50]:
df.head(20) #20 baris awal dari dataset

Unnamed: 0  Student Id Marital status  Application mode  \
0            0        6634         single               1.0   
1            1        6459         single               1.0   
2            2        2238         single              42.0   
3            3        2479         single               1.0   
4            4        4260         single              44.0   
5            5        3544        married              39.0   
6            6        4956         single              39.0   
7            7        6758         single               1.0   
8            8        2484         single               1.0   
9            9        9845         single              39.0   
10          10        3690         single               1.0   
11          11        2585         single               1.0   
12          12        5241         single               1.0   
13          13        2972         single               1.0   
14          14        2003        married              39.0   
15          15        3007         single               7.0   
16          16        9857         single              17.0   
17          17        3327         single               NaN   
18          18        5235         single              17.0   
19          19        3969         single               1.0   

    Application order  Course Daytime/evening attendance\t  \
0                 3.0  9773.0                          day   
1                 1.0     NaN                          day   
2                 1.0  9500.0                          day   
3                 1.0     NaN                          day   
4                 1.0     NaN                          day   
5                 1.0  9853.0                          day   
6                 1.0  9085.0                          day   
7                 4.0  9670.0                          day   
8                 1.0  9773.0                          day   
9                 1.0  9670.0                          day   
10                1.0  9670.0                          day   
11                1.0  9070.0                          day   
12                1.0   171.0                          day   
13                3.0  9500.0                          day   
14                1.0  8014.0                          NaN   
15                1.0  9119.0                          day   
16                1.0  9119.0                          day   
17                1.0  9003.0                          day   
18                1.0   171.0                          day   
19                1.0  9085.0                          day   

    Previous qualification  Previous qualification (grade) Nacionality  ...  \
0                      1.0                             NaN  Portuguese  ...   
1                      1.0                           136.0  Portuguese  ...   
2                      1.0                           120.0  Portuguese  ...   
3                      1.0                           141.0  Portuguese  ...   
4                     39.0                           120.0     Spanish  ...   
5                      1.0                             NaN  Portuguese  ...   
6                      1.0                           133.1         NaN  ...   
7                      1.0                           130.0  Portuguese  ...   
8                      1.0                           135.0  Portuguese  ...   
9                      1.0                           120.0  Portuguese  ...   
10                     1.0                           133.1  Portuguese  ...   
11                     1.0                           117.0  Portuguese  ...   
12                     1.0                           162.0  Portuguese  ...   
13                     1.0                           135.0  Portuguese  ...   
14                     NaN                           120.0  Portuguese  ...   
15                     1.0                           140.8  Portuguese  ...   
16                     1.0        

In [51]:
df.tail(20) # 20 baris terakhir dataset

Unnamed: 0  Student Id Marital status  Application mode  \
3076        3076        6757         single              17.0   
3077        3077        5750         single               1.0   
3078        3078        2599         single               1.0   
3079        3079        3249         single              17.0   
3080        3080        3366       divorced              39.0   
3081        3081        3462         single              39.0   
3082        3082        9194         single              17.0   
3083        3083        5574         single               1.0   
3084        3084        8311         single               1.0   
3085        3085        2304         single              17.0   
3086        3086        6615         single              43.0   
3087        3087        3759         single               1.0   
3088        3088        8085         single              51.0   
3089        3089        1573         single              17.0   
3090        3090        9737         single              51.0   
3091        3091        6176         single              44.0   
3092        3092        8064         single               1.0   
3093        3093        2103         single               1.0   
3094        3094        8629         single              39.0   
3095        3095        6614         single              17.0   

      Application order  Course Daytime/evening attendance\t  \
3076                NaN  9254.0                          day   
3077                2.0  9500.0                          day   
3078                1.0  9085.0                          day   
3079                1.0  9238.0                          day   
3080                1.0  9003.0                          day   
3081                1.0   171.0                          day   
3082                5.0  9147.0                          day   
3083                1.0  9500.0                          day   
3084                2.0  9773.0                          day   
3085                1.0  9238.0                          day   
3086                1.0  9556.0                          day   
3087                1.0   171.0                          day   
3088                1.0  9670.0                          day   
3089                2.0   171.0                          day   
3090                1.0  9070.0                          day   
3091                1.0  9085.0                          day   
3092                5.0  9238.0                          day   
3093                2.0  9500.0                          day   
3094                1.0  9500.0                          day   
3095                1.0  9670.0                          day   

      Previous qualification  Previous qualification (grade) Nacionality  ...  \
3076                     1.0                           122.0  Portuguese  ...   
3077                     1.0                           140.0  Portuguese  ...   
3078                     1.0                           131.0  Portuguese  ...   
3079                     1.0                           120.0  Portuguese  ...   
3080                    12.0                           133.1  Portuguese  ...   
3081                     1.0                           147.0  Portuguese  ...   
3082                     1.0                           131.0  Portuguese  ...   
3083                     1.0                           140.0  Portuguese  ...   
3084                     1.0                           130.0  Portuguese  ...   
3085                     NaN                           139.0  Portuguese  ...   
3086                     1.0                           149.0  Portuguese  ...   
3087                     3.0                           116.0  Portuguese  ...   
3088                     1.0                           133.1  Portuguese  ...   
3089                     1.0                           142.0  Portuguese  ...   
3090                     1.0                           143.0  Portuguese  ...   

In [52]:
df.iloc[80:101] # Baris dari index 80 - 100

Unnamed: 0  Student Id Marital status  Application mode  \
80           80        1238         single              17.0   
81           81        4121         single              39.0   
82           82        2890         single               1.0   
83           83        4358         single               1.0   
84           84        4896         single               NaN   
85           85        4735         single              18.0   
86           86        1047         single               1.0   
87           87        7430         single              39.0   
88           88        8527         single               1.0   
89           89        5420         single              43.0   
90           90        5625         single              44.0   
91           91        8990            NaN               1.0   
92           92        9143        married              44.0   
93           93        1944        married               7.0   
94           94        3666         single               1.0   
95           95        6960         single              17.0   
96           96        7620         single               1.0   
97           97        6083         single              42.0   
98           98        5525         single              17.0   
99           99        6029         single              17.0   
100         100        9749         single              39.0   

     Application order  Course Daytime/evening attendance\t  \
80                 1.0  9670.0                          day   
81                 1.0  9003.0                          day   
82                 1.0   171.0                          day   
83                 1.0  9147.0                          day   
84                 1.0  9991.0                      evening   
85                 1.0  9500.0                          day   
86                 1.0  9773.0                          day   
87                 2.0  9556.0                          day   
88                 2.0  9500.0                          day   
89                 1.0     NaN                          day   
90                 1.0  9003.0                          day   
91                 1.0  9500.0                          day   
92                 1.0  9085.0                          day   
93                 1.0  9070.0                          day   
94                 3.0  9070.0                          day   
95                 3.0  9119.0                          day   
96                 1.0  9500.0                          day   
97                 1.0  9500.0                          day   
98                 5.0  9238.0                          day   
99                 1.0  9773.0                          day   
100                1.0  9085.0                          day   

     Previous qualification  Previous qualification (grade) Nacionality  ...  \
80                      1.0                           155.0  Portuguese  ...   
81                      1.0                             NaN  Portuguese  ...   
82                      1.0                           159.0         NaN  ...   
83                      1.0                           150.0   Ukrainian  ...   
84                      1.0                           140.0  Portuguese  ...   
85                      1.0                           144.0  Portuguese  ...   
86                      1.0                           117.0  Portuguese  ...   
87                     19.0                           133.1  Portuguese  ...   
88                      1.0                           145.0  Portuguese  ...   
89                      1.0                           121.0  Portuguese  ...   
90                     39.0                           140.0  Portuguese  ...   
91                      1.0                           141.0  Portuguese  ...   
92                     39.0                           160.0  Portuguese  ...   
93                     43.0                           140.0  Portuguese  ...   
94       

# 6️. Tangani nilai yang hilang (NaN)

In [53]:
df.isnull().sum()

Unnamed: 0                                          0
Student Id                                          0
Marital status                                     84
Application mode                                  172
Application order                                 108
Course                                            152
Daytime/evening attendance\t                       80
Previous qualification                            123
Previous qualification (grade)                     85
Nacionality                                       185
Mother's qualification                            145
Father's qualification                             96
Mother's occupation                                62
Father's occupation                               144
Admission grade                                   141
Displaced                                         122
Educational special needs                         119
Debtor                                            116
Tuition fees up to date                           156
Gender                                            202
Scholarship holder                                166
Age at enrollment                                 204
International                                     101
Curricular units 1st sem (credited)               208
Curricular units 1st sem (enrolled)               143
Curricular units 1st sem (evaluations)             67
Curricular units 1st sem (approved)               184
Curricular units 1st sem (grade)                   89
Curricular units 1st sem (without evaluations)     73
Curricular units 2nd sem (credited)                61
Curricular units 2nd sem (enrolled)                99
Curricular units 2nd sem (evaluations)            207
Curricular units 2nd sem (approved)               151
Curricular units 2nd sem (grade)                   75
Curricular units 2nd sem (without evaluations)    153
Unemployment rate                                 136
Inflation rate                                    167
GDP                                                64
Target                                              0
dtype: int64

In [54]:
df.head(10) # 10 baris pertama sebelum imputasi

Unnamed: 0  Student Id Marital status  Application mode  Application order  \
0           0        6634         single               1.0                3.0   
1           1        6459         single               1.0                1.0   
2           2        2238         single              42.0                1.0   
3           3        2479         single               1.0                1.0   
4           4        4260         single              44.0                1.0   
5           5        3544        married              39.0                1.0   
6           6        4956         single              39.0                1.0   
7           7        6758         single               1.0                4.0   
8           8        2484         single               1.0                1.0   
9           9        9845         single              39.0                1.0   

   Course Daytime/evening attendance\t  Previous qualification  \
0  9773.0                          day                     1.0   
1     NaN                          day                     1.0   
2  9500.0                          day                     1.0   
3     NaN                          day                     1.0   
4     NaN                          day                    39.0   
5  9853.0                          day                     1.0   
6  9085.0                          day                     1.0   
7  9670.0                          day                     1.0   
8  9773.0                          day                     1.0   
9  9670.0                          day                     1.0   

   Previous qualification (grade) Nacionality  ...  \
0                             NaN  Portuguese  ...   
1                           136.0  Portuguese  ...   
2                           120.0  Portuguese  ...   
3                           141.0  Portuguese  ...   
4                           120.0     Spanish  ...   
5                             NaN  Portuguese  ...   
6                           133.1         NaN  ...   
7                           130.0  Portuguese  ...   
8                           135.0  Portuguese  ...   
9                           120.0  Portuguese  ...   

   Curricular units 2nd sem (credited)  Curricular units 2nd sem (enrolled)  \
0                                  0.0                                  6.0   
1                                  0.0                                  0.0   
2                                  4.0                                  8.0   
3                                  0.0                                  6.0   
4                                  1.0                                  6.0   
5                                  0.0                                  7.0   
6                                  0.0                                  6.0   
7                                  0.0                                  6.0   
8                                  0.0                                  6.0   
9                                  0.0                                  5.0   

   Curricular units 2nd sem (evaluations)  \
0                                     6.0   
1                                     0.0   
2                                     8.0   
3                                     7.0   
4                                     NaN   
5                                    10.0   
6                                    14.0   
7                                     7.0   
8                                     7.0   
9                                     8.0   

   Curricular units 2nd sem (approved)  Curricular units 2nd sem (grade)  \
0                                  6.0                         15.333333   
1                                  0.0                          0.000000   
2                                  7.0                         13.285714   
3                                  6.0                         13.142857   
4                                  3.0     

In [55]:
df.isnull().sum() # jumlah Nilai Kosong per Kolom

Unnamed: 0                                          0
Student Id                                          0
Marital status                                     84
Application mode                                  172
Application order                                 108
Course                                            152
Daytime/evening attendance\t                       80
Previous qualification                            123
Previous qualification (grade)                     85
Nacionality                                       185
Mother's qualification                            145
Father's qualification                             96
Mother's occupation                                62
Father's occupation                               144
Admission grade                                   141
Displaced                                         122
Educational special needs                         119
Debtor                                            116
Tuition fees up to date                           156
Gender                                            202
Scholarship holder                                166
Age at enrollment                                 204
International                                     101
Curricular units 1st sem (credited)               208
Curricular units 1st sem (enrolled)               143
Curricular units 1st sem (evaluations)             67
Curricular units 1st sem (approved)               184
Curricular units 1st sem (grade)                   89
Curricular units 1st sem (without evaluations)     73
Curricular units 2nd sem (credited)                61
Curricular units 2nd sem (enrolled)                99
Curricular units 2nd sem (evaluations)            207
Curricular units 2nd sem (approved)               151
Curricular units 2nd sem (grade)                   75
Curricular units 2nd sem (without evaluations)    153
Unemployment rate                                 136
Inflation rate                                    167
GDP                                                64
Target                                              0
dtype: int64

# miceforest

In [56]:
# ambil hanya kolom numerik
numeric_cols = df.select_dtypes(include=['number']).columns
df_numeric = df[numeric_cols].copy()

# bersihkan semua karakter non-alfanumerik di nama kolom
df_numeric.columns = [
    re.sub(r'[^A-Za-z0-9_]+', '_', c) for c in df_numeric.columns
]

# jalankan miceforest
kernel = mf.ImputationKernel(
    data=df_numeric,
    random_state=42
)
kernel.mice(3)
df_imputed_numeric = kernel.complete_data(dataset=0)

# balikin nama kolom aslinya
df_imputed_numeric.columns = numeric_cols

# gabungkan hasil imputasi numerik ke dataset asli
df_imputed = df.copy()
df_imputed[numeric_cols] = df_imputed_numeric

print("Imputasi miceforest selesai tanpa eror:")
print(df_imputed.isnull().sum())

Imputasi miceforest selesai tanpa eror:
Unnamed: 0                                          0
Student Id                                          0
Marital status                                     84
Application mode                                    0
Application order                                   0
Course                                              0
Daytime/evening attendance\t                       80
Previous qualification                              0
Previous qualification (grade)                      0
Nacionality                                       185
Mother's qualification                              0
Father's qualification                              0
Mother's occupation                                 0
Father's occupation                                 0
Admission grade                                     0
Displaced                                           0
Educational special needs                           0
Debtor                                    

# 7. encoding terhadap kolom bertipe object menjadi numerik

In [57]:
obj_cols = df_imputed.select_dtypes(include='object').columns
if len(obj_cols) > 0:
    print("\nMelakukan Encoding pada Kolom:", list(obj_cols))
    le = LabelEncoder()
    for col in obj_cols:
        df_imputed[col] = le.fit_transform(df_imputed[col].astype(str))
else:
    print("\ntidak ada kolom bertipe object untuk di-encode.")


Melakukan Encoding pada Kolom: ['Marital status', 'Daytime/evening attendance\t', 'Nacionality', 'Target']


# SimpelImputer

In [58]:
from sklearn.impute import SimpleImputer

# Cek apakah ada kolom kategorikal
cat_cols = df_imputed.select_dtypes(include='object').columns

if len(cat_cols) == 0:
    print("Tidak ada kolom kategorikal yang perlu diimputasi.")
else:
    print("Kolom kategorikal yang akan diimputasi:", list(cat_cols))
    imputer_cat = SimpleImputer(strategy='most_frequent')
    df_imputed[cat_cols] = imputer_cat.fit_transform(df_imputed[cat_cols])
    print("\nImputasi kategorikal selesai.")
    print(df_imputed.isnull().sum())


Tidak ada kolom kategorikal yang perlu diimputasi.


In [59]:
df_imputed.isnull().sum() # cek apakah masih ada missing value atau tidak

Unnamed: 0                                        0
Student Id                                        0
Marital status                                    0
Application mode                                  0
Application order                                 0
Course                                            0
Daytime/evening attendance\t                      0
Previous qualification                            0
Previous qualification (grade)                    0
Nacionality                                       0
Mother's qualification                            0
Father's qualification                            0
Mother's occupation                               0
Father's occupation                               0
Admission grade                                   0
Displaced                                         0
Educational special needs                         0
Debtor                                            0
Tuition fees up to date                           0
Gender                                            0
Scholarship holder                                0
Age at enrollment                                 0
International                                     0
Curricular units 1st sem (credited)               0
Curricular units 1st sem (enrolled)               0
Curricular units 1st sem (evaluations)            0
Curricular units 1st sem (approved)               0
Curricular units 1st sem (grade)                  0
Curricular units 1st sem (without evaluations)    0
Curricular units 2nd sem (credited)               0
Curricular units 2nd sem (enrolled)               0
Curricular units 2nd sem (evaluations)            0
Curricular units 2nd sem (approved)               0
Curricular units 2nd sem (grade)                  0
Curricular units 2nd sem (without evaluations)    0
Unemployment rate                                 0
Inflation rate                                    0
GDP                                               0
Target                                            0
dtype: int64

# 8️. Tampilkan kembali dataset yang telah melalui seluruh proses preprocessing (cleaning, imputasi, dan encoding)

In [65]:
df_imputed #menampilkan data final

Unnamed: 0  Student Id  Marital status  Application mode  \
0              0        6634               5               1.0   
1              1        6459               5               1.0   
2              2        2238               5              42.0   
3              3        2479               5               1.0   
4              4        4260               5              44.0   
...          ...         ...             ...               ...   
3091        3091        6176               5              44.0   
3092        3092        8064               5               1.0   
3093        3093        2103               5               1.0   
3094        3094        8629               5              39.0   
3095        3095        6614               5              17.0   

      Application order  Course  Daytime/evening attendance\t  \
0                   3.0  9773.0                             0   
1                   1.0   171.0                             0   
2                   1.0  9500.0                             0   
3                   1.0  9070.0                             0   
4                   1.0  9085.0                             0   
...                 ...     ...                           ...   
3091                1.0  9085.0                             0   
3092                5.0  9238.0                             0   
3093                2.0  9500.0                             0   
3094                1.0  9500.0                             0   
3095                1.0  9670.0                             0   

      Previous qualification  Previous qualification (grade)  Nacionality  \
0                        1.0                           130.0           13   
1                        1.0                           136.0           13   
2                        1.0                           120.0           13   
3                        1.0                           141.0           13   
4                       39.0                           120.0           17   
...                      ...                             ...          ...   
3091                     1.0                           140.0           13   
3092                     1.0                           126.0           13   
3093                     1.0                           141.0           13   
3094                     4.0                           150.0           13   
3095                     1.0                           118.0           13   

      ...  Curricular units 2nd sem (credited)  \
0     ...                                  0.0   
1     ...                                  0.0   
2     ...                                  4.0   
3     ...                                  0.0   
4     ...                                  1.0   
...   ...                                  ...   
3091  ...                                  1.0   
3092  ...                                  0.0   
3093  ...                                  0.0   
3094  ...                                  2.0   
3095  ...                                  0.0   

      Curricular units 2nd sem (enrolled)  \
0                                     6.0   
1                                     0.0   
2                                     8.0   
3                                     6.0   
4                                     6.0   
...                                   ...   
3091                                  6.0   
3092                                  6.0   
3093                                  8.0   
3094                                  8.0   
3095                                  6.0   

      Curricular units 2nd sem (evaluations)  \
0                                        6.0   
1                                        0.0   
2                                        8.0   
3                                        7.0   
4                                       15.0   
...                                      ...   
3091                         